[API] Add conditional entity lock check with interface and trait

shinde-rahul opened the pull request:
- Introduced LockableModelInterface to ensure only API models implementing locking are checked.
- Extracted locking logic to a reusable trait that reuses FormModel::isLocked().
- Updated API controller to conditionally apply locking only to LockableModelInterface models.
- Added test coverage for non-lockable entities (e.g. Lead) to ensure editing is allowed.


| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | ❌
| New feature/enhancement? (use the a.x branch)      | ✔️
| Deprecations?                          | ❌
| BC breaks? (use the c.x branch)        | ❌
| Automated tests included?              | ✔️
| Related user documentation PR URL      | mautic/user-documentation#... <!-- required for new features -->
| Related developer documentation PR URL | mautic/developer-documentation-new#257
| Issue(s) addressed                     | Fixes #... <!-- prefix each issue number with "Fixes #", no need to create an issue if none exists, explain below instead -->


## Description


This PR improves API behavior by enforcing entity locking where supported — specifically, for entities like Email — to prevent concurrent edits and to communicate locked status to API consumers.

### User Stories
As a user, I want my changes in the editor not to be overwritten via the API if I’m currently editing the entity.

As an API user, I want to know if someone else is editing the email, so I can respond appropriately (e.g. show a warning or avoid update attempts).

### What’s Included
- Introduced LockableModelInterface under ApiBundle\Model to isolate API-specific locking behavior.
- Added LockableModelTrait to reuse existing FormModel::isLocked() logic in a clean, consistent way.
- Updated Common API controller logic to check for isLocked() only when supported by the entity model.
- Functional test coverage added for:
   - Email entity: Locked state returns 409 Conflict with translated error message.
   - Lead entity: Proves non-lockable entities are still editable via the API.

### Why

This ensures that lockable entities in the UI behave consistently at the API level, avoiding data loss or silent overwrites, and making it possible for API clients to respond gracefully when an entity is locked.


---
### 📋 Steps to test this PR:

A video showing how I tested this,

https://github.com/user-attachments/assets/ef376a51-6771-4c05-9fc0-224d94e55d35

#### For Verify Locked Email Entity Returns API Conflict (409)

1. Open this PR on Gitpod or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))
2. Create a new Email
3. As an Admin user, edit the Email to simulate locking
4. Using Postman or curl, attempt to update the Email via the API:
```
curl --location --request PATCH '/api/emails/{id}/edit' \
--header 'Content-Type: application/json' \
--header 'Authorization: Basic {access_token}' \
--data '{
  "name": "Updated Email"
}
```
Use a different user than "Admin User"
5. Observe the API response with Status Code: 409 Conflict and a body
```
{
    "errors": [
        {
            "code": 409,
            "message": "{ email name} is currently checked out by {user name} (on {date}).",
            "details": {
                "checkedOutBy": "{user name}",
                "checkedOut": "{date}"
            }
        }
    ]
}
```

#### Verify Contact Entity Can Be Edited Even If “Locked”

1. Open this PR on Gitpod or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))
2. Create a new Contact with email address
3. As an Admin user, edit the newly created Contact to simulate locking
4. Using Postman or curl, attempt to update the Contact via the API:
```
curl --location --request PATCH '/api/emails/{id}/edit' \
--header 'Content-Type: application/json' \
--header 'Authorization: Basic {access_token}' \
--data '{
  "firstname": "Updated first name"
}
```
Use a different user than "Admin User"
5. Observe the API response 
   - Status Code: 200 OK
   - Response includes updated contact details, showing that the API does not enforce lock for this entity.

shinde-rahul committed:
Add API-level locking support for entities with LockableModelInterface

- Introduced LockableModelInterface to ensure only API models implementing locking are checked.
- Extracted locking logic to a reusable trait that reuses FormModel::isLocked().
- Updated API controller to conditionally apply locking only to LockableModelInterface models.
- Added test coverage for non-lockable entities (e.g. Lead) to ensure editing is allowed.
- Implemented a data provider to simplify future test additions for non-lockable entities.

RCheesley commented:
@shinde-rahul I don't see a dev docs / user docs PR for this, please ensure that you write this up as it'll be important to understand how to work with locked resources.

Bastian2718 commented:
If I sent a Patch Request via Insomnia with a new name for the email the email got changed even if it's locked. I created another User and that User can't edit the email but the Patch Request can.

Bastian2718 reviewed (approved):
Ahhhh. Now it works as expected. For email it is not editable and for contacts it is. Nice work!

RCheesley committed:
Merge branch '7.x' into feature/api-entity-lock-support

RCheesley commented:
Great job, thanks @shinde-rahul and thanks @Bastian2718 for testing, too!

shinde-rahul committed:
Merge branch '7.x' into feature/api-entity-lock-support

abhisekmazumdar commented on app/bundles/ApiBundle/Tests/Functional/Controller/CommonApiControllerTest.php:
```diff
@@ -0,0 +1,98 @@
+        $expectedMessage = $translator->trans('mautic.api.error.entity.locked', [
+            '%name%' => $email->getName(),
+            '%user%' => $email->getCheckedOutByUser(),
+            '%date%' => $email->getCheckedOut()->format('F j, Y'),
```
Is the format diff ok then what is in the controller.
```
'%date%' => $date->format($this->coreParametersHelper->get('date_format_dateonly')),
'%time%' => $date->format($this->coreParametersHelper->get('date_format_timeonly')),
```
I'm not sure about the format. Just can you cross check it ?
I see the test are passing though

escopecz replied:
@shinde-rahul there seem to be no response to this.

escopecz replied:
@shinde-rahul ?

shinde-rahul replied:
I missed it, will include this suggestion. Thanks @abhisekmazumdar and @escopecz for patience.

abhisekmazumdar reviewed (approved):
The test video was super helpful to understand this quickly and test it my self.

shinde-rahul committed:
Merge branch '7.x' into feature/api-entity-lock-support

kuzmany reviewed (approved):
Theoretically, we should also integrate it with `editEntitiesAction`, but we can proceed as is 👍

shinde-rahul committed:
Merge branch '7.x' into feature/api-entity-lock-support

shinde-rahul committed:
Update the testcase to use the date and time format from the same souce not custom one

shinde-rahul committed:
Fix API lock response to use model-defined entity name getter

shinde-rahul committed:
enforce entity lock check in processForm

Moves API lock conflict handling into processForm so batch edit returns per-entity 409; adds functional coverage for batch edit locked email.
